// angular-split/bundles/angular-split.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define('angular-split', ['exports', '@angular/core', 'rxjs', 'rxjs/operators', '@angular/common', '@angular/platform-browser'], factory) :
    (factory((global['angular-split'] = {}),global.ng.core,global.rxjs,global.rxjs.operators,global.ng.common,global.ng.platformBrowser));
}(this, (function (exports,core,rxjs,operators,common,platformBrowser) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    /**
     * @param {?} event
     * @return {?}
     */
    function getPointFromEvent(event) {
        // TouchEvent
        if ((( /** @type {?} */(event))).touches !== undefined && (( /** @type {?} */(event))).touches.length > 0) {
            return {
                x: (( /** @type {?} */(event))).touches[0].clientX,
                y: (( /** @type {?} */(event))).touches[0].clientY,
            };
        }
        // MouseEvent
        else if ((( /** @type {?} */(event))).clientX !== undefined && (( /** @type {?} */(event))).clientY !== undefined) {
            return {
                x: (( /** @type {?} */(event))).clientX,
                y: (( /** @type {?} */(event))).clientY,
            };
        }
        return null;
    }
    /**
     * @param {?} elRef
     * @param {?} direction
     * @return {?}
     */
    function getPixelSize(elRef, direction) {
        return elRef.nativeElement[(direction === 'horizontal') ? 'offsetWidth' : 'offsetHeight'];
    }
    /**
     * @param {?} v
     * @return {?}
     */
    function getInputBoolean(v) {
        return (typeof (v) === 'boolean') ? v : (v === 'false' ? false : true);
    }
    /**
     * @param {?} total
     * @return {?}
     */
    function isValidTotalSize(total) {
        return total > .999 && total < 1.001;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    /**
     * angular-split
     *
     * Areas size are set in percentage of the split container.
     * Gutters size are set in pixels.
     *
     * So we set css 'flex-basis' property like this (where 0 <= area.size <= 1):
     *  calc( { area.size * 100 }% - { area.size * nbGutter * gutterSize }px );
     *
     * Examples with 3 visible areas and 2 gutters:
     *
     * |                     10px                   10px                                  |
     * |---------------------[]---------------------[]------------------------------------|
     * |  calc(20% - 4px)          calc(20% - 4px)              calc(60% - 12px)          |
     *
     *
     * |                          10px                        10px                        |
     * |--------------------------[]--------------------------[]--------------------------|
     * |  calc(33.33% - 6.667px)      calc(33.33% - 6.667px)      calc(33.33% - 6.667px)  |
     *
     *
     * |10px                                                  10px                        |
     * |[]----------------------------------------------------[]--------------------------|
     * |0                 calc(66.66% - 13.333px)                  calc(33%% - 6.667px)   |
     *
     *
     *  10px 10px                                                                         |
     * |[][]------------------------------------------------------------------------------|
     * |0 0                               calc(100% - 20px)                               |
     *
     */
    var SplitComponent = /** @class */ (function () {
        function SplitComponent(ngZone, elRef, cdRef, renderer) {
            this.ngZone = ngZone;
            this.elRef = elRef;
            this.cdRef = cdRef;
            this.renderer = renderer;
            this._direction = 'horizontal';
            ////
            this._gutterSize = 11;
            ////
            this._useTransition = false;
            ////
            this._disabled = false;
            ////
            this._dir = 'ltr';
            this.dragProgressSubject = new rxjs.Subject();
            this.dragProgress$ = this.dragProgressSubject.asObservable();
            ////
            this.isDragging = false;
            this.currentGutterNum = 0;
            this.startPoint = null;
            this.endPoint = null;
            this.displayedAreas = [];
            this.hidedAreas = [];
            this.dragListeners = [];
            this.dragStartValues = {
                sizePixelContainer: 0,
                sizePixelA: 0,
                sizePixelB: 0,
                sizePercentA: 0,
                sizePercentB: 0,
            };
            // To force adding default class, could be override by user @Input() or not
            this.direction = this._direction;
        }
        Object.defineProperty(SplitComponent.prototype, "direction", {
            get: /**
             * @return {?}
             */ function () {
                return this._direction;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                this._direction = (v === 'vertical') ? 'vertical' : 'horizontal';
                this.renderer.addClass(this.elRef.nativeElement, "is-" + this._direction);
                this.renderer.removeClass(this.elRef.nativeElement, "is-" + ((this._direction === 'vertical') ? 'horizontal' : 'vertical'));
                this.build(false, false);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitComponent.prototype, "gutterSize", {
            get: /**
             * @return {?}
             */ function () {
                return this._gutterSize;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                v = Number(v);
                this._gutterSize = (!isNaN(v) && v > 0) ? v : 11;
                this.build(false, false);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitComponent.prototype, "useTransition", {
            get: /**
             * @return {?}
             */ function () {
                return this._useTransition;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                this._useTransition = getInputBoolean(v);
                if (this._useTransition)
                    this.renderer.addClass(this.elRef.nativeElement, 'is-transition');
                else
                    this.renderer.removeClass(this.elRef.nativeElement, 'is-transition');
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitComponent.prototype, "disabled", {
            get: /**
             * @return {?}
             */ function () {
                return this._disabled;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                this._disabled = getInputBoolean(v);
                if (this._disabled)
                    this.renderer.addClass(this.elRef.nativeElement, 'is-disabled');
                else
                    this.renderer.removeClass(this.elRef.nativeElement, 'is-disabled');
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitComponent.prototype, "dir", {
            get: /**
             * @return {?}
             */ function () {
                return this._dir;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                v = (v === 'rtl') ? 'rtl' : 'ltr';
                this._dir = v;
                this.renderer.setAttribute(this.elRef.nativeElement, 'dir', this._dir);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitComponent.prototype, "dragStart", {
            get: /**
             * @return {?}
             */ function () {
                var _this = this;
                return new rxjs.Observable(function (subscriber) { return _this.dragStartSubscriber = subscriber; });
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitComponent.prototype, "dragEnd", {
            get: /**
             * @return {?}
             */ function () {
                var _this = this;
                return new rxjs.Observable(function (subscriber) { return _this.dragEndSubscriber = subscriber; });
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitComponent.prototype, "gutterClick", {
            get: /**
             * @return {?}
             */ function () {
                var _this = this;
                return new rxjs.Observable(function (subscriber) { return _this.gutterClickSubscriber = subscriber; });
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitComponent.prototype, "transitionEnd", {
            get: /**
             * @return {?}
             */ function () {
                var _this = this;
                return new rxjs.Observable(function (subscriber) { return _this.transitionEndSubscriber = subscriber; }).pipe(operators.debounceTime(20));
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        SplitComponent.prototype.ngAfterViewInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.ngZone.runOutsideAngular(function () {
                    // To avoid transition at first rendering
                    setTimeout(function () { return _this.renderer.addClass(_this.elRef.nativeElement, 'is-init'); });
                });
            };
        /**
         * @return {?}
         */
        SplitComponent.prototype.getNbGutters = /**
         * @return {?}
         */
            function () {
                return (this.displayedAreas.length === 0) ? 0 : this.displayedAreas.length - 1;
            };
        /**
         * @param {?} component
         * @return {?}
         */
        SplitComponent.prototype.addArea = /**
         * @param {?} component
         * @return {?}
         */
            function (component) {
                /** @type {?} */
                var newArea = {
                    component: component,
                    order: 0,
                    size: 0,
                };
                if (component.visible === true) {
                    this.displayedAreas.push(newArea);
                    this.build(true, true);
                }
                else {
                    this.hidedAreas.push(newArea);
                }
            };
        /**
         * @param {?} component
         * @return {?}
         */
        SplitComponent.prototype.removeArea = /**
         * @param {?} component
         * @return {?}
         */
            function (component) {
                if (this.displayedAreas.some(function (a) { return a.component === component; })) {
                    /** @type {?} */
                    var area = this.displayedAreas.find(function (a) { return a.component === component; });
                    this.displayedAreas.splice(this.displayedAreas.indexOf(area), 1);
                    this.build(true, true);
                }
                else if (this.hidedAreas.some(function (a) { return a.component === component; })) {
                    /** @type {?} */
                    var area = this.hidedAreas.find(function (a) { return a.component === component; });
                    this.hidedAreas.splice(this.hidedAreas.indexOf(area), 1);
                }
            };
        /**
         * @param {?} component
         * @param {?} resetOrders
         * @param {?} resetSizes
         * @return {?}
         */
        SplitComponent.prototype.updateArea = /**
         * @param {?} component
         * @param {?} resetOrders
         * @param {?} resetSizes
         * @return {?}
         */
            function (component, resetOrders, resetSizes) {
                // Only refresh if area is displayed (No need to check inside 'hidedAreas')
                /** @type {?} */
                var area = this.displayedAreas.find(function (a) { return a.component === component; });
                if (!area) {
                    return;
                }
                this.build(resetOrders, resetSizes);
            };
        /**
         * @param {?} component
         * @return {?}
         */
        SplitComponent.prototype.showArea = /**
         * @param {?} component
         * @return {?}
         */
            function (component) {
                var _a;
                /** @type {?} */
                var area = this.hidedAreas.find(function (a) { return a.component === component; });
                if (!area) {
                    return;
                }
                /** @type {?} */
                var areas = this.hidedAreas.splice(this.hidedAreas.indexOf(area), 1);
                (_a = this.displayedAreas).push.apply(_a, __spread(areas));
                this.build(true, true);
            };
        /**
         * @param {?} comp
         * @return {?}
         */
        SplitComponent.prototype.hideArea = /**
         * @param {?} comp
         * @return {?}
         */
            function (comp) {
                var _a;
                /** @type {?} */
                var area = this.displayedAreas.find(function (a) { return a.component === comp; });
                if (!area) {
                    return;
                }
                /** @type {?} */
                var areas = this.displayedAreas.splice(this.displayedAreas.indexOf(area), 1);
                areas.forEach(function (area) {
                    area.order = 0;
                    area.size = 0;
                });
                (_a = this.hidedAreas).push.apply(_a, __spread(areas));
                this.build(true, true);
            };
        /**
         * @return {?}
         */
        SplitComponent.prototype.getVisibleAreaSizes = /**
         * @return {?}
         */
            function () {
                return this.displayedAreas.map(function (a) { return a.size * 100; });
            };
        /**
         * @param {?} sizes
         * @return {?}
         */
        SplitComponent.prototype.setVisibleAreaSizes = /**
         * @param {?} sizes
         * @return {?}
         */
            function (sizes) {
                if (sizes.length !== this.displayedAreas.length) {
                    return false;
                }
                sizes = sizes.map(function (s) { return s / 100; });
                /** @type {?} */
                var total = sizes.reduce(function (total, v) { return total + v; }, 0);
                if (!isValidTotalSize(total)) {
                    return false;
                }
                this.displayedAreas.forEach(function (area, i) {
                    // @ts-ignore
                    area.component._size = sizes[i];
                });
                this.build(false, true);
                return true;
            };
        /**
         * @param {?} resetOrders
         * @param {?} resetSizes
         * @return {?}
         */
        SplitComponent.prototype.build = /**
         * @param {?} resetOrders
         * @param {?} resetSizes
         * @return {?}
         */
            function (resetOrders, resetSizes) {
                var _this = this;
                this.stopDragging();
                // ¤ AREAS ORDER
                if (resetOrders === true) {
                    // If user provided 'order' for each area, use it to sort them.
                    if (this.displayedAreas.every(function (a) { return a.component.order !== null; })) {
                        this.displayedAreas.sort(function (a, b) { return (( /** @type {?} */(a.component.order))) - (( /** @type {?} */(b.component.order))); });
                    }
                    // Then set real order with multiples of 2, numbers between will be used by gutters.
                    this.displayedAreas.forEach(function (area, i) {
                        area.order = i * 2;
                        area.component.setStyleOrder(area.order);
                    });
                }
                // ¤ AREAS SIZE PERCENT
                if (resetSizes === true) {
                    /** @type {?} */
                    var totalUserSize = ( /** @type {?} */(this.displayedAreas.reduce(function (total, s) { return s.component.size ? total + s.component.size : total; }, 0)));
                    // If user provided 'size' for each area and total == 1, use it.
                    if (this.displayedAreas.every(function (a) { return a.component.size !== null; }) && isValidTotalSize(totalUserSize)) {
                        this.displayedAreas.forEach(function (area) {
                            area.size = ( /** @type {?} */(area.component.size));
                        });
                    }
                    // Else set equal sizes for all areas.
                    else {
                        /** @type {?} */
                        var size_1 = 1 / this.displayedAreas.length;
                        this.displayedAreas.forEach(function (area) {
                            area.size = size_1;
                        });
                    }
                }
                // ¤ 
                // If some real area sizes are less than gutterSize, 
                // set them to zero and dispatch size to others.
                /** @type {?} */
                var percentToDispatch = 0;
                // Get container pixel size
                /** @type {?} */
                var containerSizePixel = getPixelSize(this.elRef, this.direction);
                this.displayedAreas.forEach(function (area) {
                    if (area.size * containerSizePixel < _this.gutterSize) {
                        percentToDispatch += area.size;
                        area.size = 0;
                    }
                });
                if (percentToDispatch > 0 && this.displayedAreas.length > 0) {
                    /** @type {?} */
                    var nbAreasNotZero = this.displayedAreas.filter(function (a) { return a.size !== 0; }).length;
                    if (nbAreasNotZero > 0) {
                        /** @type {?} */
                        var percentToAdd_1 = percentToDispatch / nbAreasNotZero;
                        this.displayedAreas.filter(function (a) { return a.size !== 0; }).forEach(function (area) {
                            area.size += percentToAdd_1;
                        });
                    }
                    // All area sizes (container percentage) are less than guterSize,
                    // It means containerSize < ngGutters * gutterSize
                    else {
                        this.displayedAreas[this.displayedAreas.length - 1].size = 1;
                    }
                }
                this.refreshStyleSizes();
                this.cdRef.markForCheck();
            };
        /**
         * @return {?}
         */
        SplitComponent.prototype.refreshStyleSizes = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var sumGutterSize = this.getNbGutters() * this.gutterSize;
                this.displayedAreas.forEach(function (area) {
                    area.component.setStyleFlexbasis("calc( " + area.size * 100 + "% - " + area.size * sumGutterSize + "px )");
                });
            };
        /**
         * @param {?} event
         * @param {?} gutterNum
         * @return {?}
         */
        SplitComponent.prototype.clickGutter = /**
         * @param {?} event
         * @param {?} gutterNum
         * @return {?}
         */
            function (event, gutterNum) {
                event.preventDefault();
                event.stopPropagation();
                if (this.startPoint && this.startPoint.x === event.clientX && this.startPoint.y === event.clientY) {
                    this.currentGutterNum = gutterNum;
                    this.notify('click');
                }
            };
        /**
         * @param {?} event
         * @param {?} gutterOrder
         * @param {?} gutterNum
         * @return {?}
         */
        SplitComponent.prototype.startDragging = /**
         * @param {?} event
         * @param {?} gutterOrder
         * @param {?} gutterNum
         * @return {?}
         */
            function (event, gutterOrder, gutterNum) {
                var _this = this;
                event.preventDefault();
                event.stopPropagation();
                this.startPoint = getPointFromEvent(event);
                if (!this.startPoint || this.disabled) {
                    return;
                }
                /** @type {?} */
                var areaA = this.displayedAreas.find(function (a) { return a.order === gutterOrder - 1; });
                /** @type {?} */
                var areaB = this.displayedAreas.find(function (a) { return a.order === gutterOrder + 1; });
                if (!areaA || !areaB) {
                    return;
                }
                this.dragStartValues.sizePixelContainer = getPixelSize(this.elRef, this.direction);
                this.dragStartValues.sizePixelA = getPixelSize(areaA.component.elRef, this.direction);
                this.dragStartValues.sizePixelB = getPixelSize(areaB.component.elRef, this.direction);
                this.dragStartValues.sizePercentA = areaA.size;
                this.dragStartValues.sizePercentB = areaB.size;
                this.currentGutterNum = gutterNum;
                this.ngZone.runOutsideAngular(function () {
                    _this.dragListeners.push(_this.renderer.listen('document', 'mouseup', _this.stopDragging.bind(_this)));
                    _this.dragListeners.push(_this.renderer.listen('document', 'touchend', _this.stopDragging.bind(_this)));
                    _this.dragListeners.push(_this.renderer.listen('document', 'touchcancel', _this.stopDragging.bind(_this)));
                    _this.dragListeners.push(_this.renderer.listen('document', 'mousemove', function (e) { return _this.dragEvent(e, areaA, areaB); }));
                    _this.dragListeners.push(_this.renderer.listen('document', 'touchmove', function (e) { return _this.dragEvent(e, areaA, areaB); }));
                });
                areaA.component.lockEvents();
                areaB.component.lockEvents();
                this.isDragging = true;
                this.renderer.addClass(this.elRef.nativeElement, 'is-dragging');
                this.renderer.addClass(this.gutterEls.toArray()[this.currentGutterNum - 1].nativeElement, 'is-dragged');
                this.notify('start');
            };
        /**
         * @param {?} event
         * @param {?} areaA
         * @param {?} areaB
         * @return {?}
         */
        SplitComponent.prototype.dragEvent = /**
         * @param {?} event
         * @param {?} areaA
         * @param {?} areaB
         * @return {?}
         */
            function (event, areaA, areaB) {
                event.preventDefault();
                event.stopPropagation();
                if (!this.isDragging) {
                    return;
                }
                this.endPoint = getPointFromEvent(event);
                if (!this.endPoint) {
                    return;
                }
                // ¤ AREAS SIZE PIXEL
                /** @type {?} */
                var offsetPixel = (this.direction === 'horizontal') ? (this.startPoint.x - this.endPoint.x) : (this.startPoint.y - this.endPoint.y);
                if (this.dir === 'rtl') {
                    offsetPixel = -offsetPixel;
                }
                /** @type {?} */
                var newSizePixelA = this.dragStartValues.sizePixelA - offsetPixel;
                /** @type {?} */
                var newSizePixelB = this.dragStartValues.sizePixelB + offsetPixel;
                if (newSizePixelA < this.gutterSize && newSizePixelB < this.gutterSize) {
                    // WTF.. get out of here!
                    return;
                }
                else if (newSizePixelA < this.gutterSize) {
                    newSizePixelB += newSizePixelA;
                    newSizePixelA = 0;
                }
                else if (newSizePixelB < this.gutterSize) {
                    newSizePixelA += newSizePixelB;
                    newSizePixelB = 0;
                }
                // ¤ AREAS SIZE PERCENT
                if (newSizePixelA === 0) {
                    areaB.size += areaA.size;
                    areaA.size = 0;
                }
                else if (newSizePixelB === 0) {
                    areaA.size += areaB.size;
                    areaB.size = 0;
                }
                else {
                    // NEW_PERCENT = START_PERCENT / START_PIXEL * NEW_PIXEL;
                    if (this.dragStartValues.sizePercentA === 0) {
                        areaB.size = this.dragStartValues.sizePercentB / this.dragStartValues.sizePixelB * newSizePixelB;
                        areaA.size = this.dragStartValues.sizePercentB - areaB.size;
                    }
                    else if (this.dragStartValues.sizePercentB === 0) {
                        areaA.size = this.dragStartValues.sizePercentA / this.dragStartValues.sizePixelA * newSizePixelA;
                        areaB.size = this.dragStartValues.sizePercentA - areaA.size;
                    }
                    else {
                        areaA.size = this.dragStartValues.sizePercentA / this.dragStartValues.sizePixelA * newSizePixelA;
                        areaB.size = (this.dragStartValues.sizePercentA + this.dragStartValues.sizePercentB) - areaA.size;
                    }
                }
                this.refreshStyleSizes();
                // If moved from starting point, notify progress
                if (this.startPoint.x !== this.endPoint.x || this.startPoint.y !== this.endPoint.y) {
                    this.notify('progress');
                }
            };
        /**
         * @param {?=} event
         * @return {?}
         */
        SplitComponent.prototype.stopDragging = /**
         * @param {?=} event
         * @return {?}
         */
            function (event) {
                var _this = this;
                if (event) {
                    event.preventDefault();
                    event.stopPropagation();
                }
                if (this.isDragging === false) {
                    return;
                }
                this.displayedAreas.forEach(function (area) {
                    area.component.unlockEvents();
                });
                while (this.dragListeners.length > 0) {
                    /** @type {?} */
                    var fct = this.dragListeners.pop();
                    if (fct) {
                        fct();
                    }
                }
                // If moved from starting point, notify end
                if (event && this.endPoint && (this.startPoint.x !== this.endPoint.x || this.startPoint.y !== this.endPoint.y)) {
                    this.notify('end');
                }
                this.isDragging = false;
                this.renderer.removeClass(this.elRef.nativeElement, 'is-dragging');
                this.renderer.removeClass(this.gutterEls.toArray()[this.currentGutterNum - 1].nativeElement, 'is-dragged');
                // Needed to let (click)="clickGutter(...)" event run and verify if mouse moved or not
                this.ngZone.runOutsideAngular(function () {
                    setTimeout(function () {
                        _this.startPoint = null;
                        _this.endPoint = null;
                    });
                });
            };
        /**
         * @param {?} type
         * @return {?}
         */
        SplitComponent.prototype.notify = /**
         * @param {?} type
         * @return {?}
         */
            function (type) {
                var _this = this;
                /** @type {?} */
                var sizes = this.displayedAreas.map(function (a) { return a.size * 100; });
                if (type === 'start') {
                    if (this.dragStartSubscriber) {
                        this.ngZone.run(function () { return _this.dragStartSubscriber.next({ gutterNum: _this.currentGutterNum, sizes: sizes }); });
                    }
                }
                else if (type === 'end') {
                    if (this.dragEndSubscriber) {
                        this.ngZone.run(function () { return _this.dragEndSubscriber.next({ gutterNum: _this.currentGutterNum, sizes: sizes }); });
                    }
                }
                else if (type === 'click') {
                    if (this.gutterClickSubscriber) {
                        this.ngZone.run(function () { return _this.gutterClickSubscriber.next({ gutterNum: _this.currentGutterNum, sizes: sizes }); });
                    }
                }
                else if (type === 'transitionEnd') {
                    if (this.transitionEndSubscriber) {
                        this.ngZone.run(function () { return _this.transitionEndSubscriber.next(sizes); });
                    }
                }
                else if (type === 'progress') {
                    // Stay outside zone to allow users do what they want about change detection mechanism.
                    this.dragProgressSubject.next({ gutterNum: this.currentGutterNum, sizes: sizes });
                }
            };
        /**
         * @return {?}
         */
        SplitComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.stopDragging();
            };
        SplitComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'as-split',
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        template: "\n        <ng-content></ng-content>\n        <ng-template ngFor [ngForOf]=\"displayedAreas\" let-index=\"index\" let-last=\"last\">\n            <div *ngIf=\"last === false\" \n                 #gutterEls\n                 class=\"as-split-gutter\"\n                 [style.flex-basis.px]=\"gutterSize\"\n                 [style.order]=\"index*2+1\"\n                 (as-split-undetected.click)=\"clickGutter($event, index+1)\"\n                 (as-split-undetected.mousedown)=\"startDragging($event, index*2+1, index+1)\"\n                 (as-split-undetected.touchstart)=\"startDragging($event, index*2+1, index+1)\">\n                <div class=\"as-split-gutter-icon\"></div>\n            </div>\n        </ng-template>",
                        styles: [":host{display:flex;flex-wrap:nowrap;justify-content:flex-start;align-items:stretch;overflow:hidden;width:100%;height:100%}:host>.as-split-gutter{flex-grow:0;flex-shrink:0;background-color:#eee;display:flex;align-items:center;justify-content:center}:host>.as-split-gutter>.as-split-gutter-icon{width:100%;height:100%;background-position:center center;background-repeat:no-repeat}:host ::ng-deep>.as-split-area{flex-grow:0;flex-shrink:0;overflow-x:hidden;overflow-y:auto}:host ::ng-deep>.as-split-area.is-hided{flex-basis:0!important;overflow-x:hidden;overflow-y:hidden}:host.is-horizontal{flex-direction:row}:host.is-horizontal>.as-split-gutter{flex-direction:row;cursor:col-resize;height:100%}:host.is-horizontal>.as-split-gutter>.as-split-gutter-icon{background-image:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAeCAYAAADkftS9AAAAIklEQVQoU2M4c+bMfxAGAgYYmwGrIIiDjrELjpo5aiZeMwF+yNnOs5KSvgAAAABJRU5ErkJggg==)}:host.is-horizontal ::ng-deep>.as-split-area{height:100%}:host.is-vertical{flex-direction:column}:host.is-vertical>.as-split-gutter{flex-direction:column;cursor:row-resize;width:100%}:host.is-vertical>.as-split-gutter .as-split-gutter-icon{background-image:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAB4AAAAFCAMAAABl/6zIAAAABlBMVEUAAADMzMzIT8AyAAAAAXRSTlMAQObYZgAAABRJREFUeAFjYGRkwIMJSeMHlBkOABP7AEGzSuPKAAAAAElFTkSuQmCC)}:host.is-vertical ::ng-deep>.as-split-area{width:100%}:host.is-vertical ::ng-deep>.as-split-area.is-hided{max-width:0}:host.is-disabled>.as-split-gutter{cursor:default}:host.is-disabled>.as-split-gutter .as-split-gutter-icon{background-image:url(\"\")}:host.is-transition.is-init:not(.is-dragging) ::ng-deep>.as-split-area,:host.is-transition.is-init:not(.is-dragging)>.as-split-gutter{transition:flex-basis .3s}"]
                    }] }
        ];
        /** @nocollapse */
        SplitComponent.ctorParameters = function () {
            return [
                { type: core.NgZone },
                { type: core.ElementRef },
                { type: core.ChangeDetectorRef },
                { type: core.Renderer2 }
            ];
        };
        SplitComponent.propDecorators = {
            direction: [{ type: core.Input }],
            gutterSize: [{ type: core.Input }],
            useTransition: [{ type: core.Input }],
            disabled: [{ type: core.Input }],
            dir: [{ type: core.Input }],
            dragStart: [{ type: core.Output }],
            dragEnd: [{ type: core.Output }],
            gutterClick: [{ type: core.Output }],
            transitionEnd: [{ type: core.Output }],
            gutterEls: [{ type: core.ViewChildren, args: ['gutterEls',] }]
        };
        return SplitComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var SplitAreaDirective = /** @class */ (function () {
        function SplitAreaDirective(ngZone, elRef, renderer, split) {
            this.ngZone = ngZone;
            this.elRef = elRef;
            this.renderer = renderer;
            this.split = split;
            this._order = null;
            ////
            this._size = null;
            ////
            this._visible = true;
            this.lockListeners = [];
            this.renderer.addClass(this.elRef.nativeElement, 'as-split-area');
        }
        Object.defineProperty(SplitAreaDirective.prototype, "order", {
            get: /**
             * @return {?}
             */ function () {
                return this._order;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                v = Number(v);
                this._order = !isNaN(v) ? v : null;
                this.split.updateArea(this, true, false);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitAreaDirective.prototype, "size", {
            get: /**
             * @return {?}
             */ function () {
                return this._size;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                v = Number(v);
                this._size = (!isNaN(v) && v >= 0 && v <= 100) ? (v / 100) : null;
                this.split.updateArea(this, false, true);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SplitAreaDirective.prototype, "visible", {
            get: /**
             * @return {?}
             */ function () {
                return this._visible;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                this._visible = getInputBoolean(v);
                if (this._visible) {
                    this.split.showArea(this);
                    this.renderer.removeClass(this.elRef.nativeElement, 'is-hided');
                }
                else {
                    this.split.hideArea(this);
                    this.renderer.addClass(this.elRef.nativeElement, 'is-hided');
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        SplitAreaDirective.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.split.addArea(this);
                this.ngZone.runOutsideAngular(function () {
                    _this.transitionListener = _this.renderer.listen(_this.elRef.nativeElement, 'transitionend', function (event) {
                        // Limit only flex-basis transition to trigger the event
                        if (event.propertyName === 'flex-basis') {
                            _this.split.notify('transitionEnd');
                        }
                    });
                });
            };
        /**
         * @param {?} value
         * @return {?}
         */
        SplitAreaDirective.prototype.setStyleOrder = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                this.renderer.setStyle(this.elRef.nativeElement, 'order', value);
            };
        /**
         * @param {?} value
         * @return {?}
         */
        SplitAreaDirective.prototype.setStyleFlexbasis = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                this.renderer.setStyle(this.elRef.nativeElement, 'flex-basis', value);
            };
        /**
         * @return {?}
         */
        SplitAreaDirective.prototype.lockEvents = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.ngZone.runOutsideAngular(function () {
                    _this.lockListeners.push(_this.renderer.listen(_this.elRef.nativeElement, 'selectstart', function (e) { return false; }));
                    _this.lockListeners.push(_this.renderer.listen(_this.elRef.nativeElement, 'dragstart', function (e) { return false; }));
                });
            };
        /**
         * @return {?}
         */
        SplitAreaDirective.prototype.unlockEvents = /**
         * @return {?}
         */
            function () {
                while (this.lockListeners.length > 0) {
                    /** @type {?} */
                    var fct = this.lockListeners.pop();
                    if (fct) {
                        fct();
                    }
                }
            };
        /**
         * @return {?}
         */
        SplitAreaDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.unlockEvents();
                if (this.transitionListener) {
                    this.transitionListener();
                }
                this.split.removeArea(this);
            };
        SplitAreaDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'as-split-area, [as-split-area]'
                    },] }
        ];
        /** @nocollapse */
        SplitAreaDirective.ctorParameters = function () {
            return [
                { type: core.NgZone },
                { type: core.ElementRef },
                { type: core.Renderer2 },
                { type: SplitComponent }
            ];
        };
        SplitAreaDirective.propDecorators = {
            order: [{ type: core.Input }],
            size: [{ type: core.Input }],
            visible: [{ type: core.Input }]
        };
        return SplitAreaDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    /**
     * Credit to Michael Strobel from:
     * https://github.com/kryops/ng2-events
     */
    var UndetectedEventPlugin = /** @class */ (function () {
        function UndetectedEventPlugin() {
        }
        /**
         * @param {?} eventName
         * @return {?}
         */
        UndetectedEventPlugin.prototype.supports = /**
         * @param {?} eventName
         * @return {?}
         */
            function (eventName) {
                return eventName.indexOf('as-split-undetected.') === 0;
            };
        /**
         * @param {?} element
         * @param {?} eventName
         * @param {?} handler
         * @return {?}
         */
        UndetectedEventPlugin.prototype.addEventListener = /**
         * @param {?} element
         * @param {?} eventName
         * @param {?} handler
         * @return {?}
         */
            function (element, eventName, handler) {
                var _this = this;
                /** @type {?} */
                var realEventName = eventName.slice(20);
                return this.manager.getZone().runOutsideAngular(function () { return _this.manager.addEventListener(element, realEventName, handler); });
            };
        UndetectedEventPlugin.decorators = [
            { type: core.Injectable }
        ];
        return UndetectedEventPlugin;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var AngularSplitModule = /** @class */ (function () {
        function AngularSplitModule() {
        }
        /**
         * @return {?}
         */
        AngularSplitModule.forRoot = /**
         * @return {?}
         */
            function () {
                return {
                    ngModule: AngularSplitModule,
                    providers: [{
                            provide: platformBrowser.EVENT_MANAGER_PLUGINS,
                            useClass: UndetectedEventPlugin,
                            multi: true
                        }]
                };
            };
        /**
         * @return {?}
         */
        AngularSplitModule.forChild = /**
         * @return {?}
         */
            function () {
                return {
                    ngModule: AngularSplitModule,
                    providers: []
                };
            };
        AngularSplitModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [
                            SplitComponent,
                            SplitAreaDirective,
                        ],
                        exports: [
                            SplitComponent,
                            SplitAreaDirective,
                        ]
                    },] }
        ];
        return AngularSplitModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    exports.AngularSplitModule = AngularSplitModule;
    exports.SplitComponent = SplitComponent;
    exports.SplitAreaDirective = SplitAreaDirective;
    exports.ɵa = UndetectedEventPlugin;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
